src/randomenv.cpp

```diff
@@ -0,0 +1,295 @@
+            if (n > 0) hasher.Write(fbuf, n);
+            /* not bothering with EINTR handling. */
+        } while (n == sizeof(fbuf));
+        close(f);
+    }
+}
+
+void AddPath(CSHA512& hasher, const char *path)
+{
+    struct stat sb;
+    if (stat(path, &sb) == 0) {
+        hasher.Write((const unsigned char*)path, strlen(path) + 1);
+        AddStat(hasher, sb);
+    }
+}
+#endif
+
+#if defined(__x86_64__) || defined(__amd64__) || defined(__i386__)
+void GetCPUID(uint32_t leaf, uint32_t subleaf, uint32_t& a, uint32_t& b, uint32_t& c, uint32_t& d)
+{
+    // We can't use __get_cpuid as it doesn't support subleafs.
+#ifdef __GNUC__
+    __cpuid_count(leaf, subleaf, a, b, c, d);
+#else
+    __asm__ ("cpuid" : "=a"(a), "=b"(b), "=c"(c), "=d"(d) : "0"(leaf), "2"(subleaf));
+#endif
+}
+
+void AddCPUID(CSHA512& hasher)
+{
+    uint32_t ax, bx, cx, dx;
+    GetCPUID(0, 0, ax, bx, cx, dx);
+    hasher << ax << bx << cx << dx;
+    uint32_t max = ax;
+    for (uint32_t leaf = 1; leaf < max; ++leaf) {
+        for (uint32_t subleaf = 0;; ++subleaf) {
+            GetCPUID(leaf, subleaf, ax, bx, cx, dx);
+            hasher << leaf << subleaf << ax << bx << cx << dx;
+            // Iterate over subleaves for leaf 4, 11, 13
+            if (leaf != 4 && leaf != 11 && leaf != 13) break;
+            if ((leaf == 4 || leaf == 13) && ax == 0) break;
+            if (leaf == 11 && (cx & 0xFF00) == 0) break;
+        }
+    }
+}
+#endif
+} // namespace
+
+void RandAddDynamicEnv(CSHA512& hasher)
+{
+    RandAddSeedPerfmon(hasher);
+
+#ifdef WIN32
+    FILETIME ftime;
+    GetSystemTimeAsFileTime(&ftime);
+    hasher << ftime;
+#else
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    struct timeval tv;
+    gettimeofday(&tv, nullptr);
+    hasher << ts.tv_sec << ts.tv_nsec << tv.tv_sec << tv.tv_usec;
+
+    // Current resource usage.
+    struct rusage usage;
+    if (getrusage(RUSAGE_SELF, &usage) == 0) {
+        hasher << usage.ru_utime.tv_sec << usage.ru_utime.tv_usec << usage.ru_stime.tv_sec << usage.ru_stime.tv_usec;
+        hasher << usage.ru_maxrss << usage.ru_minflt << usage.ru_majflt << usage.ru_inblock << usage.ru_oublock;
+        hasher << usage.ru_nvcsw << usage.ru_nivcsw;
+    }
+#endif
+
+}
+
+void RandAddStaticEnv(CSHA512& hasher)
+{
+    // Some compile-time static properties
+    uint32_t x;
+    x = ((CHAR_MIN < 0) << 30) + (sizeof(void *) << 16) + (sizeof(long) << 8) + sizeof(int);
+    hasher << x;
+#if defined(__GNUC__) && defined(__GNUC_MINOR__) && defined(__GNUC_PATCHLEVEL__)
+    x = (__GNUC_PATCHLEVEL__ << 16) + (__GNUC_MINOR__ << 8) + __GNUC__;
+    hasher << x;
+#endif
+#if defined(_MSC_VER)
+    x = _MSC_VER;
+    hasher << x;
+#endif
+#if defined(__STDC_VERSION__)
+    x = __STDC_VERSION__;
+    hasher << x;
+#endif
+#if defined(_XOPEN_VERSION)
+    x = _XOPEN_VERSION;
+    hasher << x;
+#endif
+
+#if defined(__x86_64__) || defined(__amd64__) || defined(__i386__)
+    AddCPUID(hasher);
+#endif
```

Comment:
To also have CPU features bits for other architectures (on Linux) you could include `getauxval(AT_HWCAP)` and `getauxval(AT_HWCAP2)`.

Reply:
Good idea, added!